lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1 +1,533 @@
+        }
+    }
+}
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole circle.
+fn circle_hashes(circle: GeoRadius, max_regions: usize) -> Vec<GeoHash> {
+    let rectangle = minimum_bounding_rectangle_for_circle(&circle);
+    let inner_hashes = rectangle_hashes(rectangle, max_regions);
+    filter_hashes_within_circle(inner_hashes, circle)
+}
+
+/// filter geo hashes for which the center is within a circle
+fn filter_hashes_within_circle(hashes: Vec<GeoHash>, circle: GeoRadius) -> Vec<GeoHash> {
+    let radius_km = circle.radius / 1000.0;
+    hashes
+        .into_iter()
+        .map(|h| {
+            let coord = decode(&h).unwrap().0;
+            let geo_point = GeoPoint {
+                lon: coord.x,
+                lat: coord.y,
+            };
+            (h, geo_point)
+        })
+        .filter(|(_, hash_point)| {
+            distance_in_km_between_coordinates(&circle.center, hash_point) <= radius_km
+        })
+        .map(|(h, _)| h)
+        .collect()
+}
+
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain the whole rectangle.
+fn rectangle_hashes(rectangle: GeoBoundingBox, max_regions: usize) -> Vec<GeoHash> {
+    let full_geo_bounding_box: FullGeoBoundingBox = (&rectangle).into();
+    let shortest_side_km = full_geo_bounding_box.shortest_side_length_in_km();
+    let possible_precisions = geohash_precisions_for_distance(shortest_side_km * 1000.0);
+    // filter precision which generates less than `max_regions`.
+    let rectangle_geohash = possible_precisions
+        .into_iter()
+        .map(|p| RectangleGeoHash::compute_from_bounding_box(&rectangle, p))
+        .take_while(|rectangle_geohash| rectangle_geohash.area_region_count() <= max_regions)
+        .last();
+    match rectangle_geohash {
+        None => Vec::new(),
+        Some(r) => r.geohash_regions(),
+    }
+}
+
+/// Finding the geo precisions fine enough to not completely wrap the `distance_in_meter` with a single tile.
+fn geohash_precisions_for_distance(distance_in_meter: f64) -> Range<usize> {
+    HASH_GRID_SIZE
+        .iter()
+        .enumerate()
+        .find(|(_, (lat_length_in_meters, _))| distance_in_meter > *lat_length_in_meters) // lat > long for all HASH_GRID_SIZE
+        .map(|(index, (_, _))| Range {
+            start: index + 1,
+            end: GEOHASH_MAX_LENGTH + 1, // The upper bound of the range (exclusive).
+        })
+        .unwrap_or_else(|| Range { start: 1, end: 1 })
+}
+
+/// A globally-average value is usually considered to be 6,371 kilometres (3,959 mi) with a 0.3% variability (±10 km).
+/// https://en.wikipedia.org/wiki/Earth_radius.
+const EARTH_RADIUS_KM: f64 = 6371.0;
+
+/// https://rust-lang-nursery.github.io/rust-cookbook/science/mathematics/trigonometry.html#distance-between-two-points-on-the-earth
+fn distance_in_km_between_coordinates(p1: &GeoPoint, p2: &GeoPoint) -> f64 {
```

Comment:
We are already using https://github.com/qdrant/qdrant/blob/02a50212e516a8601829c97b7b8facf388ad7c49/lib/segment/src/payload_storage/condition_checker.rs#L6,  let's stick to one implementation for consistency